src/utils/chunkColouring.ts

```diff
@@ -0,0 +1,83 @@
+import ExternalModule from '../ExternalModule';
+import Module from '../Module';
+import { randomUint8Array, Uint8ArrayXor } from './entryHashing';
+import { error } from './error';
+import { relative } from './path';
+
+export function assignChunkColouringHashes(
+	entryModules: Module[],
+	manualChunkModules: Record<string, Module[]>
+) {
+	let currentEntry: Module, currentEntryHash: Uint8Array;
+	let modulesVisitedForCurrentEntry: { [id: string]: boolean };
+	const handledEntryPoints: { [id: string]: boolean } = {};
+	const dynamicImports: Set<Module> = new Set();
+
+	const addCurrentEntryColourToModule = (module: Module) => {
+		if (currentEntry.chunkAlias) {
+			module.chunkAlias = currentEntry.chunkAlias;
+			module.entryPointsHash = currentEntryHash;
+		} else {
+			Uint8ArrayXor(module.entryPointsHash, currentEntryHash);
+		}
+
+		for (const dependency of module.dependencies) {
+			if (dependency instanceof ExternalModule || dependency.id in modulesVisitedForCurrentEntry) {
+				continue;
+			}
+			modulesVisitedForCurrentEntry[dependency.id] = true;
+			if (!handledEntryPoints[dependency.id] && !dependency.chunkAlias)
+				addCurrentEntryColourToModule(dependency);
+		}
+
+		for (const { resolution } of module.dynamicImports) {
+			if (
+				resolution instanceof Module &&
+				resolution.isDynamicEntryPoint &&
+				!resolution.chunkAlias
+			) {
+				dynamicImports.add(resolution);
```

Comment:
I think we need to be doing some colouring for dynamic imports to make this algorithm work properly.

Here is a test case I tried that isn't working correctly:

main.js
```
import './dep1.js';
console.log('main');
```

dep1.js
```js
import './dep2.js';
console.log('dep1');
```

dep2.js
```js
import('./dep1.js');
console.log('dep2');
```

The dynamic import back to dep1 should be informing a colouring of dep1.js being an entry point chunk effectively.

Instead I got the following output:

main.js
```js
console.log('main');
console.log('dep1');
import('./generated-chunk.js');
```

generated-chunk.js
```js
console.log('dep2');
```

The problem with this is that it has broken the execution order invariant that we require dep2 to execute statically and synchronously before dep1 and main.

Reply:
Here's the key point from #2407 that I feel may have been missed here (correct me if I'm wrong):

**it would actually be an improvement to the chunking algorithm if we could work out the common base dependencies of a given dynamic import (the intersection of the static dependency graph of all modules that use this dynamic import)**

Reply:
I am sorry but I cannot follow entirely. Using your test case, I (IMO correctly) get the following two chunks:

```js
// main.js
import './generated-chunk.js';

console.log('main');

// generated-chunk.js
Promise.resolve().then(function () { return dep1; });
console.log('dep2');

console.log('dep1');

var dep1 = /*#__PURE__*/Object.freeze({

});
```

which looks correct to me. Is there something I am missing. Adding `inlineDynamicImports`, this all gets folded into one chunk correctly. One could argue that it could have been a single chunk all along but that is another point.

I must admit I also do not understand what you mean with your key point from #2407 . Maybe you can illustrate this with some examples of the form input -> expected output?

Reply:
I tried this again and you're right it's working correctly - not sure why it wasn't working in my test previously.